File: src/utils/roleplayBlocks.ts
/**
 * 语义分块（阶段5「语义块渲染」§6.1）。
 *
 * 从自然语言正文得到确定性的 RoleplayBlock[]：
 * - 纯 `“……”` / `「……」` / `『……』` / `"……"` 行 → dialogue（可带说话人前缀）；
 * - `<thought>...</thought>` → thought；
 * - 整段 `*动作*` → narration（剥离星号，样式由 CSS 呈现）；
 * - 一行同时含对白与叙述 → mixed（保持原文，不强行补角色名）；
 * - 其余普通段落 → narration。
 *
 * 分块只影响展示：Message.content 保持可复制的原始正文（§6.2），
 * 旧消息按现有 Markdown 规则渲染（由 Message.contentRenderMode 区分）。
 *
 * 解析层是纯函数：同一输入 + phase 得到确定结果。流式阶段检测到
 * “说话人前缀 + 左引号”即产出 dialogue 且 complete=false，右引号到达后
 * 只更新正文与 complete，不改变 block 类型。
 */

export type RoleplayBlockKind = 'dialogue' | 'narration' | 'thought' | 'mixed'

export type RoleplayBlock =
  | { kind: 'dialogue'; text: string; speaker?: string; complete?: boolean }
  | { kind: 'narration'; text: string }
  | { kind: 'thought'; text: string }
  | { kind: 'mixed'; text: string }

export type RoleplayParsePhase = 'streaming' | 'final'

export interface ParseRoleplayBlocksOptions {
  phase?: RoleplayParsePhase
}

/** 配对引号：左 → 右 */
export const QUOTE_PAIRS: ReadonlyArray<readonly [string, string]> = [
  ['“', '”'],
  ['「', '」'],
  ['『', '』'],
  ['"', '"'],
]

/** 用于匹配 SPEAKER/纯对白的引号字符集（四类左/右引号；不含 ASCII 单引号，避免撇号误伤） */
const OPEN_QUOTE_CHARS = '“「『"'
const CLOSE_QUOTE_CHARS = '”」』"'

/**
 * 说话人前缀：冒号前的名称（1–8 个非空白非标点字符，覆盖中英文角色名与
 * "绿色外星人"这类指代名；带标点/空格的长前缀是叙述句，不可能是名字）。
 * 再用 NARRATION_PREFIX 排除叙述式前缀，避免"她轻声说道：/苏晚推开门："
 * 这类叙述被误当说话人。remark-roleplay 与 blocks 共用本规则。
 *
 * 右侧允许四类引号 + 可选空白后开始对白；流式阶段可能尚未闭合右引号。
 */
export const SPEAKER_QUOTE =
  new RegExp(
    `^([^${OPEN_QUOTE_CHARS}${CLOSE_QUOTE_CHARS}：:：\\s，。！？、；,.!?;'"()（）【】〔〕—…·\\-]{1,8})[：:]\\s*([${OPEN_QUOTE_CHARS}][\\s\\S]*)$`,
  )

/**
 * 叙述式前缀特征字：代词、结构助词、常见动作/说话动词。命中任一即视为叙述而非角色名。
 * 注意收录边界： transliteration 名常用字（拉/洛/诺/丽等）不得入表，避免"美洛拉"被误杀；
 * 误杀的后果只是退化为匿名对白块（可接受的降级），而漏杀会把叙述整段错标成角色。
 */
export const NARRATION_PREFIX = /[她他它您我谁的了着是在推说道问答笑喊叫哼吼念讲走站坐打握摇抬皱眨顿望听看感摆挥耸抿咬飘]/

/** 整行纯对白（四类引号成对；流式 phase 下允许未闭合） */
export const PURE_QUOTE = new RegExp(`^[${OPEN_QUOTE_CHARS}][\\s\\S]*[${CLOSE_QUOTE_CHARS}]$`)

/** 仅有左引号（流式/异常未闭合） */
const LEADING_OPEN_QUOTE = new RegExp(`^[${OPEN_QUOTE_CHARS}][\\s\\S]*$`)

/** 整行动作段（旧样式星号包裹） */
const ACTION_LINE = /^\*(.+)\*$/s

/** 引号字符（判断行内是否含对白；不含 ASCII 单引号，避免英文撇号整行误判 mixed） */
const CONTAINS_QUOTE = /[“”「」『』"]/

/** 按码点 trim（与 textMetrics 口径一致；\u3000 = 全角空格） */
function trimLine(line: string): string {
  return line.replace(/^[\s\u3000]+|[\s\u3000]+$/g, '')
}

/**
 * 从 text[0]=open 出发，找与首引号配对的 close 下标（嵌套同类引号按深度配对）。
 * 找不到返回 -1。ASCII 同字符引号按「成对游走」计数。
 */
function findMatchingCloseIndex(text: string, open: string, close: string): number {
  if (open === close) {
    let seen = 0
    for (let i = 0; i < text.length; i += 1) {
      if (text[i] === open) {
        seen += 1
        if (seen % 2 === 0) return i
      }
    }
    return -1
  }
  let depth = 0
  for (let i = 0; i < text.length; i += 1) {
    const ch = text[i]
    if (ch === open) {
      depth += 1
    } else if (ch === close) {
      depth -= 1
      if (depth === 0) return i
    }
  }
  return -1
}

/**
 * 从 text[0]=open 出发的闭合判定：
 * - `endsWithClose`：与首引号配对的 close 正好落在末尾 → 完整纯对白
 *   （嵌套 “他说：“好。”” 按深度配对；不能只看末字符，否则 `"A"叙述"B"` 会被误判）
 * - `hasInnerClose`：文中存在与首引号配对的 close，但不在末尾 → 闭合后仍有尾随（mixed）
 * - 否则：未闭合（流式 incomplete）
 */
function classifyOpenQuote(text: string): { endsWithClose: boolean; hasInnerClose: boolean } {
  if (!text || text.length < 1) return { endsWithClose: false, hasInnerClose: false }
  for (const [open, close] of QUOTE_PAIRS) {
    if (text[0] !== open) continue
    const matchEnd = findMatchingCloseIndex(text, open, close)
    if (matchEnd < 0) return { endsWithClose: false, hasInnerClose: false }
    if (matchEnd === text.length - 1) return { endsWithClose: true, hasInnerClose: true }
    return { endsWithClose: false, hasInnerClose: true }
  }
  return { endsWithClose: false, hasInnerClose: false }
}

/** 起始引号是否整段闭合 */
function isQuoteClosed(text: string): boolean {
  return classifyOpenQuote(text).endsWithClose
}

/**
 * 说话人前缀匹配（共用规则）；命中且非叙述式前缀时返回：
 * - complete: 整段为「名：纯对白」
 * - streamingOpen: 「名：」+ 未闭合左引号（流式/异常输出）
 * - 否则 null（含闭合对白后仍有尾随叙述 → 调用方走 mixed）
 */
export function matchSpeakerDialogue(
  line: string,
  _phase: RoleplayParsePhase = 'final',
): { speaker: string; dialogue: string; complete: boolean; streamingOpen: boolean } | null {
  const m = line.match(SPEAKER_QUOTE)
  if (!m) return null
  const speaker = trimLine(m[1] ?? '')
  if (!speaker || NARRATION_PREFIX.test(speaker)) return null
  const dialogue = m[2] ?? ''
  if (![...OPEN_QUOTE_CHARS].includes(dialogue[0] ?? '')) return null
  const q = classifyOpenQuote(dialogue)
  if (q.endsWithClose) {
    return { speaker, dialogue, complete: true, streamingOpen: false }
  }
  // 闭合引号后仍有文字（`“你好。”她…`）→ 不是纯对白
  if (q.hasInnerClose) return null
  // 未闭合：流式或最终异常都保持 dialogue incomplete（不吞正文）
  return { speaker, dialogue, complete: false, streamingOpen: true }
}

/** 是否整行纯对白（首尾配对为单一外层引号，或未闭合的左引号起始行） */
export function isPureDialogueLine(line: string, _phase: RoleplayParsePhase = 'final'): boolean {
  if (!line) return false
  if (!LEADING_OPEN_QUOTE.test(line) || ![...OPEN_QUOTE_CHARS].includes(line[0] ?? '')) return false
  const q = classifyOpenQuote(line)
  // 外层引号深度配对且收在末尾 → 纯对白（允许内部嵌套引号与冒号）
  if (q.endsWithClose) return true
  // 闭合后仍有文字（含 `"对白"叙述"对白"`）→ mixed
  if (q.hasInnerClose) return false
  // 未闭合：以左引号起即 incomplete 对白（流式/异常）；冒号不影响，说话人路径已单独处理
  return true
}

/** 单行 → 块 */
function classifyLine(rawLine: string, phase: RoleplayParsePhase): RoleplayBlock {
  const line = trimLine(rawLine)
  if (!line) return { kind: 'narration', text: '' }

  // 旧样式动作段：*动作* → narration（剥离星号，避免与 CSS 样式叠加）
  const action = line.match(ACTION_LINE)
  if (action) return { kind: 'narration', text: trimLine(action[1]) }

  // 说话人 + 纯对白：苏晚：“我知道。” / 叶：“走吧。”（单字角色名同样成立）
  const speaker = matchSpeakerDialogue(line, phase)
  if (speaker) {
    return {
      kind: 'dialogue',
      text: speaker.dialogue,
      speaker: speaker.speaker,
      ...(speaker.complete ? {} : { complete: false }),
    }
  }

  // 整行纯对白
  if (isPureDialogueLine(line, phase)) {
    const complete = isQuoteClosed(line)
    return {
      kind: 'dialogue',
      text: line,
      ...(complete ? {} : { complete: false }),
    }
  }

  // 对白与叙述混写 → mixed，保持原文
  if (CONTAINS_QUOTE.test(line)) return { kind: 'mixed', text: line }

  return { kind: 'narration', text: line }
}

/**
 * 将正文解析为语义块。
 * - 思考块（可跨行）优先解析为 thought；
 * - 相邻 narration 行合并为一个块（保持段落自然）；
 * - 不改写除剥离包裹星号之外的任何字符。
 *
 * @param content 原始正文
 * @param options.phase streaming：允许未闭合对白保持 dialogue；final：默认
 */
export function parseRoleplayBlocks(
  content: string,
  options: ParseRoleplayBlocksOptions = {},
): RoleplayBlock[] {
  const phase: RoleplayParsePhase = options.phase ?? 'final'
  if (!content) return []

  const blocks: RoleplayBlock[] = []
  // 先以 thought 块为分隔切分（thought 可跨行），其余按行分类
  const thoughtRe = /<thought>([\s\S]*?)<\/thought>\s*/gi
  let lastIndex = 0
  for (const match of content.matchAll(thoughtRe)) {
    const before = content.slice(lastIndex, match.index)
    appendLines(blocks, before, phase)
    const inner = trimLine(match[1] ?? '')
    if (inner) blocks.push({ kind: 'thought', text: inner })
    lastIndex = (match.index ?? 0) + match[0].length
  }
  appendLines(blocks, content.slice(lastIndex), phase)

  return blocks.filter((b) => !(b.kind === 'narration' && !b.text))
}

/** 兼容入口：最终态解析（历史调用方） */
export function buildRoleplayBlocks(content: string): RoleplayBlock[] {
  return parseRoleplayBlocks(content, { phase: 'final' })
}

/**
 * 展示层剥除对白外层引号（“…”/「…」/『…』/ "…" 成对时去掉）。
 * 只影响渲染：RoleplayBlock.text 与 message.content 保持原文，复制/搜索不受影响。
 * 未闭合时不剥，避免流式布局跳动。
 */
export function stripOuterQuotes(text: string): string {
  const t = text.trim()
  for (const [open, close] of QUOTE_PAIRS) {
    if (t.startsWith(open) && t.endsWith(close) && t.length >= 2) {
      if (open === close && t.length === 1) return text
      return t.slice(1, -1)
    }
  }
  return text
}

/**
 * mixed 段行内对白切分：引号对（含引号本身）标 quoted，其余保持原文。
 * 支持四类引号；与 classifyOpenQuote 同源深度/游走配对，避免嵌套被首个 close 截断。
 * 流式未闭合时尾部开引号保持未标记（原文）。
 */
export function splitQuoteSegments(text: string): Array<{ text: string; quoted: boolean }> {
  if (!text) return []
  const segments: Array<{ text: string; quoted: boolean }> = []
  let lastIndex = 0

  /** 扫描从 i 起能匹配的最早完整引号对；命中返回 [start, endInclusive] */
  const findQuotedSpan = (from: number): { start: number; end: number } | null => {
    for (let j = from; j < text.length; j += 1) {
      for (const [open, close] of QUOTE_PAIRS) {
        if (text[j] !== open) continue
        const end = findMatchingCloseIndex(text.slice(j), open, close)
        if (end > 0) return { start: j, end: j + end }
      }
    }
    return null
  }

  let cursor = 0
  while (cursor < text.length) {
    const span = findQuotedSpan(cursor)
    if (!span) break
    if (span.start > lastIndex) {
      segments.push({ text: text.slice(lastIndex, span.start), quoted: false })
    }
    segments.push({ text: text.slice(span.start, span.end + 1), quoted: true })
    lastIndex = span.end + 1
    cursor = lastIndex
  }
  if (lastIndex < text.length) {
    segments.push({ text: text.slice(lastIndex), quoted: false })
  }
  return segments
}

/** 把一段纯文本按行分类并追加到 blocks（同一段落内的相邻 narration 行合并；空行分隔段落） */
function appendLines(blocks: RoleplayBlock[], text: string, phase: RoleplayParsePhase): void {
  if (!text) return
  // 按空行切段：不同段落的 narration 不合并
  const paragraphs = text.split(/\n\s*\n/)
  for (const paragraph of paragraphs) {
    let lastNarration: RoleplayBlock | null = null
    for (const rawLine of paragraph.split(/\n/)) {
      const line = trimLine(rawLine)
      if (!line) continue
      const block = classifyLine(line, phase)
      const prev = blocks[blocks.length - 1]
      if (block.kind === 'narration' && prev && prev.kind === 'narration' && prev === lastNarration) {
        prev.text = `${prev.text}\n${block.text}`
      } else {
        blocks.push(block)
        if (block.kind === 'narration') lastNarration = block
      }
    }
  }
}

File: android/app/src/main/java/com/qingyu/companion/ui/components/RoleplayBlocks.kt
package com.qingyu.companion.ui.components

/**
 * 语义分块（阶段7.2，方案 §6.2）：PC `src/utils/roleplayBlocks.ts` 的 Kotlin 等价移植。
 *
 * 两端共用同一份契约样本 `shared/fixtures/roleplay-blocks.json`（权威在 PC 侧），
 * RoleplayBlocksFixtureTest 逐条比对块数量与类型，防止 TypeScript/Kotlin 规则漂移。
 *
 * 规则与 PC 一致：
 * - `<thought>...</thought>` 转 thought（可跨行）；
 * - 纯引号段（“”/「」/『』/"") 转 dialogue（可带 1-8 字说话人前缀，叙述式前缀排除）；
 * - 旧整段 `*动作*` 转 narration（剥离星号）；
 * - 同行混合内容转 mixed，按普通正文安全显示；
 * - 其余普通段落转 narration；同段相邻 narration 合并，空行分块；
 * - 只影响展示样式，不改写除剥离包裹星号之外的任何字符。
 */
sealed class RoleplayBlock {
    data class Dialogue(val text: String, val speaker: String?) : RoleplayBlock()
    class Narration(var text: String) : RoleplayBlock()
    data class Thought(val text: String) : RoleplayBlock()
    data class Mixed(val text: String) : RoleplayBlock()
}

object RoleplayBlocks {

    /**
     * JavaScript `s` 字符类的完整覆盖集（Java 的 `s` 只含 ASCII 空白，直接使用会导致
     * 全角空格 / NBSP 等 trim 行为漂移，故显式列字符类保持跨端一致）。
     */
    private const val JS_SPACE = "\\t\\n\\u000B\\u000C\\r \\u00A0\\u1680\\u2000-\\u200A\\u2028\\u2029\\u202F\\u205F\\u3000\\uFEFF"

    /** JS trim / s 视同空白的字符集（0x0B/0x0C 用 toChar 构造，规避字面控制字符） */
    private val JS_SPACE_CHARS: Set<Char> = setOf(
        '\t', '\n', 0x0B.toChar(), 0x0C.toChar(), '\r', ' ',
        '\u00A0', '\u1680', '\u2000', '\u2001', '\u2002', '\u2003',
        '\u2004', '\u2005', '\u2006', '\u2007', '\u2008', '\u2009',
        '\u200A', '\u2028', '\u2029', '\u202F', '\u205F', '\u3000',
        '\uFEFF',
    )

    /**
     * 说话人前缀 + 纯对白：冒号前 1-8 个非空白非标点字符（对齐 PC SPEAKER_QUOTE），
     * 再用 NARRATION_PREFIX 排除叙述式前缀（代词/助词/动作动词），避免
     * "她轻声说道：/苏晚推开门："被误当说话人。
     * 右侧只需左引号起始（闭合由 isSingleOuterQuote 深度判定），对齐 PC 未闭合 incomplete。
     */
    internal val SPEAKER_QUOTE =
        Regex("^([^“”「」『』：:$JS_SPACE，。！？、；,.!?;'\"()（）【】〔〕—…·\\-]{1,8})[：:][$JS_SPACE]*([“「『\"][\\s\\S]*)$")

    /** 叙述式前缀特征字（与 PC NARRATION_PREFIX 同集合）：命中任一即视为叙述而非角色名 */
    internal val NARRATION_PREFIX =
        Regex("[她他它您我谁的了着是在推说道问答笑喊叫哼吼念讲走站坐打握摇抬皱眨顿望听看感摆挥耸抿咬飘]")

    /** 整行纯对白（宽松字符类；精确判定走 isPureDialogueLine） */
    internal val PURE_QUOTE = Regex("^[“「『\"][\\s\\S]*[”」』\"]$")

    /** 整行动作段（旧样式星号包裹，等价 JS 带 s 标志的星号整行匹配） */
    private val ACTION_LINE = Regex("^\\*(.+)\\*$", setOf(RegexOption.DOT_MATCHES_ALL))

    /** 行内是否含引号字符（判断混写；不含 ASCII 单引号，避免英文撇号误伤） */
    private val CONTAINS_QUOTE = Regex("[“”「」『』\"]")

    /** 思考块（可跨行，后随空白一并吞掉） */
    private val THOUGHT_BLOCK =
        Regex("<thought>([\\s\\S]*?)</thought>[$JS_SPACE]*", RegexOption.IGNORE_CASE)

    /** 段落边界：空行 */
    private val BLANK_LINE = Regex("\\n[$JS_SPACE]*\\n")

    /**
     * 展示层剥除对白外层引号（四类成对时去掉；对齐 PC stripOuterQuotes）。
     * 只影响渲染：块 text 与消息正文保持原文，复制/搜索不受影响。
     */
    fun stripOuterQuotes(text: String): String {
        val t = trimLine(text)
        for ((open, close) in QUOTE_PAIRS) {
            if (t.startsWith(open) && t.endsWith(close) && t.length >= 2) {
                if (open == close && t.length == 1) return text
                return t.substring(1, t.length - 1)
            }
        }
        return text
    }

    /** 与 PC QUOTE_PAIRS 对齐的四类引号 */
    private val QUOTE_PAIRS: List<Pair<Char, Char>> = listOf(
        '“' to '”',
        '「' to '」',
        '『' to '』',
        '"' to '"',
    )

    /** 与首引号配对的 close 下标（嵌套按深度；open==close 按成对游走）；找不到返回 -1 */
    internal fun findMatchingCloseIndex(text: String, open: Char, close: Char): Int {
        if (open == close) {
            var seen = 0
            for (i in text.indices) {
                if (text[i] == open) {
                    seen += 1
                    if (seen % 2 == 0) return i
                }
            }
            return -1
        }
        var depth = 0
        for (i in text.indices) {
            val ch = text[i]
            if (ch == open) {
                depth += 1
            } else if (ch == close) {
                depth -= 1
                if (depth == 0) return i
            }
        }
        return -1
    }

    private fun isOpenQuote(c: Char): Boolean =
        c == '“' || c == '「' || c == '『' || c == '"'

    private fun pairedClose(open: Char): Char = when (open) {
        '“' -> '”'
        '「' -> '」'
        '『' -> '』'
        else -> open
    }

    /**
     * 整行纯对白：与首引号深度配对的 close 落在末尾，或未闭合左引号起始行。
     * 不能只看「以引号开头且以引号结尾」——“A”叙述“B” 会误判为对白块。
     */
    internal fun isPureDialogueLine(line: String): Boolean {
        if (line.isEmpty() || !isOpenQuote(line[0])) return false
        val close = pairedClose(line[0])
        val matchEnd = findMatchingCloseIndex(line, line[0], close)
        // 外层深度配对且收在末尾 → 纯对白（允许内部嵌套与冒号）
        if (matchEnd >= 0 && matchEnd == line.length - 1) return true
        // 闭合后仍有文字 → mixed
        if (matchEnd >= 0) return false
        // 未闭合：以左引号起即 incomplete 对白（冒号不影响，对齐 PC）
        return true
    }

    /** 对白段（可含内部冒号）是否为单一外层引号闭合/未闭合。对齐 PC classifyOpenQuote */
    internal fun isSingleOuterQuote(dialogue: String): Boolean {
        if (dialogue.isEmpty() || !isOpenQuote(dialogue[0])) return false
        val close = pairedClose(dialogue[0])
        val matchEnd = findMatchingCloseIndex(dialogue, dialogue[0], close)
        if (matchEnd < 0) return true // 未闭合
        return matchEnd == dialogue.length - 1
    }

    /**
     * 说话人 + 纯对白（对齐 PC matchSpeakerDialogue）：返回 null 表示非说话人对白结构。
     * 供 Markdown 兼容路径与 classifyLine 共用，避免旧路径继续用松散 PURE_QUOTE。
     */
    fun matchSpeakerDialogue(line: String): Pair<String, String>? {
        val m = SPEAKER_QUOTE.matchEntire(line) ?: return null
        val speaker = trimLine(m.groupValues[1])
        if (speaker.isEmpty() || NARRATION_PREFIX.containsMatchIn(speaker)) return null
        val dialogue = m.groupValues[2]
        if (!isSingleOuterQuote(dialogue)) return null
        return speaker to dialogue
    }

    private fun isJsSpace(c: Char): Boolean = JS_SPACE_CHARS.contains(c)

    /** 按码点 trim（对齐 TS trimLine：JS 空白 + 全角空格） */
    private fun trimLine(line: String): String =
        line.dropWhile { isJsSpace(it) }.reversed().dropWhile { isJsSpace(it) }.reversed()

    /** 单行归类（对齐 TS classifyLine） */
    private fun classifyLine(rawLine: String): RoleplayBlock {
        val line = trimLine(rawLine)
        if (line.isEmpty()) return RoleplayBlock.Narration("")

        // 旧样式动作段：*动作* 转 narration（剥离星号，避免与样式叠加）
        ACTION_LINE.matchEntire(line)?.let { action ->
            return RoleplayBlock.Narration(trimLine(action.groupValues[1]))
        }

        // 说话人 + 纯对白：苏晚：“我知道。”（单字角色名同样成立；叙述式前缀排除）
        matchSpeakerDialogue(line)?.let { (speaker, dialogue) ->
            return RoleplayBlock.Dialogue(dialogue, speaker = speaker)
        }

        // 整行纯对白（深度配对，排除 “A”叙述“B” 混写）
        if (isPureDialogueLine(line)) return RoleplayBlock.Dialogue(line, null)

        // 对白与叙述混写 转 mixed，保持原文
        if (CONTAINS_QUOTE.containsMatchIn(line)) return RoleplayBlock.Mixed(line)

        return RoleplayBlock.Narration(line)
    }

    /**
     * 将正文解析为语义块（对齐 TS buildRoleplayBlocks）：
     * 思考块优先切分，其余按行分类；只影响展示，不改写正文。
     */
    fun build(content: String?): List<RoleplayBlock> {
        if (content.isNullOrEmpty()) return emptyList()

        val blocks = mutableListOf<RoleplayBlock>()
        var lastIndex = 0
        for (match in THOUGHT_BLOCK.findAll(content)) {
            appendLines(blocks, content.substring(lastIndex, match.range.first))
            val inner = trimLine(match.groupValues[1])
            if (inner.isNotEmpty()) blocks.add(RoleplayBlock.Thought(inner))
            lastIndex = match.range.last + 1
        }
        appendLines(blocks, content.substring(lastIndex))

        return blocks.filterNot { it is RoleplayBlock.Narration && it.text.isEmpty() }
    }

    /** 纯文本按行分类并追加（同段相邻 narration 合并；空行分隔段落；对齐 TS appendLines） */
    private fun appendLines(blocks: MutableList<RoleplayBlock>, text: String) {
        if (text.isEmpty()) return
        val paragraphs = BLANK_LINE.split(text)
        for (paragraph in paragraphs) {
            var lastNarration: RoleplayBlock.Narration? = null
            for (rawLine in paragraph.split('\n')) {
                val line = trimLine(rawLine)
                if (line.isEmpty()) continue
                val block = classifyLine(line)
                val prev = blocks.lastOrNull()
                if (block is RoleplayBlock.Narration && prev != null && prev === lastNarration) {
                    prev.text = prev.text + "\n" + block.text
                } else {
                    blocks.add(block)
                    if (block is RoleplayBlock.Narration) lastNarration = block
                }
            }
        }
    }
}
